Machine_Learning_cheatsheet (#141)
AlexsLemonade/training-modules: module_cheatsheets/Machine_Learning_Cheatsheet.md

# Machine Learning Cheatsheet

#### The tables below consist of valuable functions or commands that will help you through this module. 
##### Each table represents a different library/tool and its corresponding commands. 
> Please note that these tables are not intended to tell you all the information you need to know about each command. 
> 
> The hyperlinks found in each piece of code will take you to the documentation for further information on the usage of each command. 
 
<br>
<br>

### AnnotationDbi 
Read the `AnnotationDbi` package vignette [**here**](http://bioconductor.org/packages/release/bioc/vignettes/AnnotationDbi/inst/doc/IntroToAnnotationPackages.pdf).

| Library/Package                                  | Piece of Code                               | What it's called                               | What it does                                                                                                                                   |
|-----------------------------------------|---------------------------------------------|-------------------------|------------------------------------------------------------------------|

> | `AnnotationDbi`        | [`keytypes()`](https://www.rdocumentation.org/packages/AnnotationDbi/versions/1.44.0/topics/AnnotationDb-objects)      | Keytypes     | Returns a character vector of column names (keytypes)                                               |

Author: I am using 'type of gene identifiers' here. Should I specify that in the notebook this would be 'ENSEMBL' or should I leave it to match the format of the below explanation for 'Mapped IDs'?

> | `AnnotationDbi`        | [`keytypes()`](https://www.rdocumentation.org/packages/AnnotationDbi/versions/1.44.0/topics/AnnotationDb-objects)      | Keytypes     | Returns a character vector of column names (keytypes)                                               |

Reviewer: Can you push the changes? I will take a look.

> | `AnnotationDbi`        | [`keytypes()`](https://www.rdocumentation.org/packages/AnnotationDbi/versions/1.44.0/topics/AnnotationDb-objects)      | Keytypes     | Returns a character vector of column names (keytypes)                                               |

Author: Sure

> | `AnnotationDbi`        | [`keytypes()`](https://www.rdocumentation.org/packages/AnnotationDbi/versions/1.44.0/topics/AnnotationDb-objects)      | Keytypes     | Returns a character vector of column names/keytypes (eg., type of gene identifiers, which was "ENSEMBL" in the notebook)                                              |

Reviewer: ```suggestion
| `AnnotationDbi`        | [`keytypes()`](https://www.rdocumentation.org/packages/AnnotationDbi/versions/1.44.0/topics/AnnotationDb-objects)      | Keytypes     | Returns a character vector of column names/keytypes (eg., type of gene identifiers) available in an `AnnotationDbi` package.                                           |
```

Author: Okay, looks good. I assume after I make this change I'll be ready to merge, correct?

Reviewer: Yep! Provided everything else has been addressed at this point (didn't look at that, trusting you).

> | `AnnotationDbi`       | [`mapIDs()`](https://www.rdocumentation.org/packages/AnnotationDbi/versions/1.44.0/topics/AnnotationDb-objects) | Mapped IDs       | Extracts the mapped ids for a set of keys of a specific keytype                                    |                                  

Reviewer: ```suggestion
| `AnnotationDbi`       | [`mapIDs()`](https://www.rdocumentation.org/packages/AnnotationDbi/versions/1.44.0/topics/AnnotationDb-objects) | Mapped IDs       | Extracts the mapped ids for a set of keys (e.g., gene identifiers) of a specific keytype                                    |                                  
```
Would you say that's fair?

| Base `R`               | [`any()`](https://www.rdocumentation.org/packages/base/versions/3.5.1/topics/any)             | Any      | Checks to see if at least one of the elements are `TRUE` when given a logical vector                                      |
| Base `R`               | [`is.na()`](https://www.rdocumentation.org/packages/base/versions/3.5.1/topics/NA)                | Is it "Not Available"            | Returns a logical vector where `TRUE` represents an instance of a missing value  |
| Base `R`               | [`cor()`](https://www.rdocumentation.org/packages/stats/versions/3.6.1/topics/cor)        | Correlation     | Computes correlation between samples using a specified correlation method, and returns a correlation matrix        |
| Base `R`               | [`table()`](https://www.rdocumentation.org/packages/base/versions/3.5.1/topics/table)    | Create Table       | Creates a contingency table of counts for each combination of factor levels               |

> | Base `R`               | [`cbind()`](https://www.rdocumentation.org/packages/base/versions/3.5.1/topics/cbind)                | Cbind                          | Combines R objects using specified columns or rows                                         |

Reviewer: `cbind` will always be columns.
```suggestion
| Base `R`               | [`cbind()`](https://www.rdocumentation.org/packages/base/versions/3.5.1/topics/cbind)                | Cbind                          | Combines vectors, matrices, or data.frames by columns                                         |
```

|-----------------------------------------|---------------------------------------------|-------------------------|------------------------------------------------------------------------|
| `PLIER`               | `combinePaths()`         | Combine Pathways                 | Combines the pathway data obtained from `PLIER` and returns the result as a matrix                                                            |
| `PLIER`                | `commonRows()`           | Common Rows                      | Determines the rows (genes) that are common to the specified data matrices and returns them as a character vector                             |
| `PLIER`                | `rowNorm()`              | Row Normalize                    | Normalizes each row (gene) by z-scoring the expression values                                                                                 |
| `PLIER`                | `num.pc()`               | Number of Principal Components   | Returns the number of significant principal components                                                                                        |

> | `PLIER`                | `PLIER()`                | Main PLIER Function              | Learns the correlated patterns of expression in the data, and returns a U matrix representing the information extracted at the pathway level  |

Reviewer: It returns a list with 14 (I think?) elements. For the cheatsheet, I would keep it pretty simple.

<div style="page-break-after: always;"></div>

### `data.table`, `reshape2`, `RColorBrewer`, `ConsensusClusterPlus`, `ggsignif`
Documentation for each of these packages can be accessed by clicking the package name in the table below.

| Library/Package                                  | Piece of Code                               | What it's called                               | What it does                                                                                                                                   |
|-----------------------------------------|---------------------------------------------|-------------------------|------------------------------------------------------------------------|
| [`data.table`](https://www.rdocumentation.org/packages/data.table/versions/1.11.8)       | [`fread()`](https://www.rdocumentation.org/packages/data.table/versions/1.11.8/topics/fread)    | F read           | Reads in data faster than base R                                                                         |

> | [`reshape2`](https://www.rdocumentation.org/packages/reshape2/versions/1.4.3)             | [`melt()`](https://www.rdocumentation.org/packages/reshape2/versions/1.4.3/topics/melt)         | Melt             | Converts an object into a data.frame                                                                           |

Reviewer: ```suggestion
| [`reshape2`](https://www.rdocumentation.org/packages/reshape2/versions/1.4.3)             | [`melt()`](https://www.rdocumentation.org/packages/reshape2/versions/1.4.3/topics/melt)         | Melt             | Converts an object into a data.frame in 'long' format                                                                           |
```

| [`RColorBrewer`](https://www.rdocumentation.org/packages/RColorBrewer/versions/1.1-2)         | [`brewer.pal()`](https://www.rdocumentation.org/packages/RColorBrewer/versions/1.1-2/topics/RColorBrewer)           | Brewer Pal       | Makes specified color palettes from ColorBrewer available in the R environment             |
| [`ConsensusClusterPlus`](https://www.rdocumentation.org/packages/ConsensusClusterPlus/versions/1.46.0) | [`ConsensusClusterPlus()`](https://www.rdocumentation.org/packages/ConsensusClusterPlus/versions/1.36.0/topics/ConsensusClusterPlus) | Consensus Clustering             | Finds the consensus across multiple runs of the clustering algorithm                              |

> | [`ggsignif`](https://www.rdocumentation.org/packages/ggsignif/versions/0.4.0)             | [`geom_signif()`](https://www.rdocumentation.org/packages/ggsignif/versions/0.4.0/topics/stat_signif)    | Create Significance Layer   | Computes the significance of the difference between groups and adds this annotation to the plot        |

Reviewer: ```suggestion
| [`ggsignif`](https://www.rdocumentation.org/packages/ggsignif/versions/0.4.0)             | [`geom_signif()`](https://www.rdocumentation.org/packages/ggsignif/versions/0.4.0/topics/stat_signif)    | Create Significance Layer   | Adds significance information to the plot. It can be used to run statistical tests and display the significance information from those tests. We use it differently, in a way that gives us more control, in the notebook.       |
```
